lib/client/ipam_block_reader_writer.go

```diff
@@ -301,32 +301,66 @@ func blockGenerator(pool cnet.IPNet) func() *cnet.IPNet {
 // block from the given pool.  When there are no blocks left,
 // the it returns nil.
 func randomBlockGenerator(pool cnet.IPNet) func() *cnet.IPNet {
+
 	// Determine the IP type to use.
 	version := getIPVersion(cnet.IP{pool.IP})
 	baseIP := cnet.IP{pool.IP}
 
 	// Determine the number of blocks within this pool.
 	ones, size := pool.Mask.Size()
 	prefixLen := size - ones
-	numBlocks := int(math.Exp2(float64(prefixLen))) / blockSize
+	numIP := new(big.Int).Exp(big.NewInt(2), big.NewInt(int64(prefixLen)), nil)
+	numBlocks := new(big.Int)
+	numBlocks.Div(numIP, big.NewInt(blockSize))
 
-	// Generate a randomly ordered slice of block indexes.
+	// Start at a random offset index
 	source := rand.NewSource(time.Now().UnixNano())
 	randm := rand.New(source)
-	idxs := randm.Perm(numBlocks)
-	i := 0
+
+	// initialIndex keeps track of the random starting point
+	initialIndex := new(big.Int)
+	initialIndex.Rand(randm, numBlocks)
+
+	// i keeps track of current index while walking the blocks in a pool
+	i := initialIndex
+
+	// numReturned keeps track of number of blocks returned
+	numReturned := big.NewInt(0)
+
+	// numLeft keeps track of how many blocks are left to assign
+	numLeft := new(big.Int)
+
+	// numDiff = numBlocks - i
+	numDiff := new(big.Int)
+
 	return func() *cnet.IPNet {
-		// Pop a block index off of the indexes slice.
-		if len(idxs) != 0 {
-			i, idxs = idxs[len(idxs)-1], idxs[:len(idxs)-1]
+		inc := i.Mul(i, big.NewInt(blockSize))
+		ip := incrementIP(baseIP, inc)
+		ipnet := net.IPNet{ip.IP, version.BlockPrefixMask}
+
+		numLeft.Sub(numBlocks, numReturned)
+		numDiff.Sub(numBlocks, i)
+
+		if numDiff.Int64() <= 1 { //Logic: `i >= numBlocks - 1`
```

Comment:
I don't think there is a guarantee that numDiff fits within an int64.  

We have 2^128 / 2^26 = 2^122 possible IPv6 blocks.  The largest difference in block indexes well exceeds max(int64).


I think we need to be careful to perform all the comparisons using the `big.Int.Cmp` method.

So I think that would be:

```
if numDiff.Cmp(big.NewInt(1)) <= 0
```